test/samlRequest.spec.ts

```diff
@@ -0,0 +1,127 @@
+import { SAML } from "../src/saml";
```

Comment:
Would a test or two to check to make sure that the incoming data is of `object` type be reasonable?

Reply:
Or rather, that an incoming type other than `object` will fail to construct instead with a sensible error instead of a compiler-generated error?

Reply:
I agree to check the type and it should be an object and throw type error with the message `samlExtensions should be Object`

Reply:
Done, Please review it again. 

Also one more thing is remaining [here](https://github.com/node-saml/node-saml/pull/11#discussion_r686789096).

>Just to be thorough, this comment indicates that you'll do another PR for the LogoutRequest version of Extensions. Does that mean that we should be using a more descriptive name instead of samlExtensions so there isn't a future name collision for that next PR?


I would like to suggest `authnSAMLRequestExtensions` instead of `samlExtensions` field, @cjbarth  @srd90 let me know your thoughts

Reply:
@kdhttps whats stopping you at implementing logout extensions in this same PR. If it would be implemented in this PR we would not have to try to see into the future what should have been taken into consideration when logout request extensions is introduced so that there would not have been any unneccesary breaking chances due late introduction of logout part.

Reply:
@srd90 , agreed but I found that it is not really an `extensions` element in logoutRequest. Could you please check again https://docs.oasis-open.org/security/saml/v2.0/saml-schema-protocol-2.0.xsd? Maybe I missed something, please guide!

Reply:
You missed that `LogoutRequestType` extends `RequestAbstractType`. `RequestAbstractType` type defines extensions element (see [this logout request related comment](https://github.com/node-saml/node-saml/pull/11#discussion_r671817007)).

Situation is same with `AuthnRequestType` (see [this authn request related comment](https://github.com/node-saml/node-saml/pull/11#discussion_r671007653)).

Some random side notes:

You have said that you have some customer who wants to pass parameters per request and furthermore that IdP is suomifi. suomifi document https://palveluhallinta.suomi.fi/en/tuki/artikkelit/59116c3014bbb10001966f70 says that language code is expected in AuthnRequest (see "Table 4" from that document) and in LogoutRequest (see "table 20" from that document).

About general usability of current impementation (from developer point of view). Assume that you have to pass e.g. 3 different request specific parameters and that each of them could have 3 different values and that each parameter's value is determined at the time authentication request is about to be issued. At the moment developer has two options:
* instantiate 3x3x2 (2 comes from authn+logout) different passport-saml instances and select the one which has correct combination request paramer values in passport-saml's configuration options because at the moment this pull request allows configuration of request parameters only at passport-saml stack instantion time.
* instantiate new passport-saml stack instance for each authentication and logout request with correct set of request specific parameters.

Alternative could to allow setting request parameres in authenticate function call also (more about this in [this comment](https://github.com/node-saml/node-saml/pull/11#issuecomment-881025810) at paragraph starting with sentence "About enabling different extensions element content per request. ...")

Reply:
Hello @srd90,
Thank you for the information.

> You missed that LogoutRequestType extends RequestAbstractType. RequestAbstractType type defines extensions element (see this logout request related comment).

Gotcha, I will work on it. I will take `samlAuthnRequestExtensions` parameter name for SAML Authn Request and `samlLogoutRequestExtensions` parameter name for SAML Logout Request.

>You have said that you have some customer who wants to pass parameters per request and furthermore that IdP is suomifi.

No, For now just a one-time config but as per your observation, it looks like others may need different extensions as per request. I would like to cover this feature in the new PR. For now, let's add only the basics functionality.

Reply:
I agree, it seems wise that we add functionality in bit-sized pieces as people need it. This seems like a base that we can work with without any breaking changes to add the new feature. Is that your estimation too @kdhttps , that adding these additional features would result in no breaking changes?

Reply:
Sorry not getting your point. I can't see any breaking changes, Both features are optionally configurable.